src/main/scala/com/fulcrumgenomics/umi/UmiConsensusCaller.scala

```diff
@@ -0,0 +1,277 @@
+    val oldRgs = in.getReadGroups.toSeq
+    def collapse(f: SAMReadGroupRecord => String): String = oldRgs.map(f).filter(_ != null).toSet.toList match {
+      case Nil      => null
+      case x :: Nil => x
+      case xs       => xs.mkString(",")
+    }
+
+    val rg = new SAMReadGroupRecord(readGroupId)
+    rg.setDescription(s"Consensus reads generated from ${oldRgs.size} input read groups.")
+    rg.setLibrary         (collapse(_.getLibrary))
+    rg.setSample          (collapse(_.getSample))
+    rg.setPlatform        (collapse(_.getPlatform))
+    rg.setPlatformUnit    (collapse(_.getPlatformUnit))
+    rg.setSequencingCenter(collapse(_.getSequencingCenter))
+
+    val outHeader = new SAMFileHeader
+    outHeader.addReadGroup(rg)
+    outHeader.setSortOrder(sortOrder.getOrElse(SortOrder.unsorted))
+    outHeader.setGroupOrder(GroupOrder.query)
+    outHeader
+  }
+}
+
+
+/**
+  * A trait that can be mixed in by any consensus caller that works at the read level,
+  * mapping incoming SAMRecords into consensus SAMRecords.
+  *
+  * @tparam C Internally, the type of lightweight conesnsus read that is used prior to
+  *           rebuilding SAMRecords.
+  */
+trait UmiConsensusCaller[C <: ConsensusRead] {
+  import com.fulcrumgenomics.umi.UmiConsensusCaller.ReadType._
+
+  // vars to track how many reads meet various fates
+  private var _totalReads: Long = 0
+  private var _totalFiltered: Long = 0
+  private var _filteredForMinorityAlignment: Long = 0
+
+  /** Returns the total number of input reads examined by the consensus caller so far. */
+  def totalReads: Long = _totalReads
+
+  /**
+    * Returns the number of raw reads filtered out due to there being insufficient reads present
+    * to build the necessary set of consensus reads.
+    */
+  def readsFilteredInsufficientSupport: Long = this._totalFiltered - this._filteredForMinorityAlignment
+
+  /** Returns the number of raw reads filtered out because their alignment disagreed with the majority alignment of
+    * all raw reads for the same source molecule
+    */
+  def readsFilteredMinorityAlignment: Long = _filteredForMinorityAlignment
+
+  /** Records that the supplied records were rejected, and not used to build a consensus read. */
+  protected def rejectRecords(recs: Traversable[SAMRecord]) : Unit = {
+    this._totalFiltered += recs.size
+  }
+
+  /** A RG.ID to apply to all generated reads. */
+  protected def readGroupId: String
+
+  /** A prefix to use on all read names.  If None then a suitable prefix will be synthesized. */
+  protected val readNamePrefix: String
+
+  /**
+    * Needs to be implemented to to return a value from a SAMRecord that represents the unit of
+    * grouping, e.g. the MI tag for vanilla UMI data and the MI tag minus the /?? suffix for
+    * duplex data.
+    * @param rec a SAMRecord
+    * @return an identified for the source molecule
+    */
+  protected[umi] def sourceMoleculeId(rec: SAMRecord): String
```

Comment:
Protected so sub-classes can see it to implement it.  Package so it can be invoked from tests.